command/server.go

```diff
@@ -1719,6 +1719,77 @@ func (c *ServerCommand) Run(args []string) int {
 				c.logger.Info(fmt.Sprintf("Wrote stacktrace to: %s", f.Name()))
 				f.Close()
 			}
+
+			// We can only get pprof outputs via the API but sometimes Vault can get
+			// into a state where it cannot process requests so we can get pprof outputs
+			// via SIGUSR2.
+			if os.Getenv("VAULT_PPROF_WRITE_TO_FILE") != "" {
+				dir := ""
+				path := os.Getenv("VAULT_PPROF_FILE_PATH")
+				if path != "" {
+					if _, err := os.Stat(path); err != nil {
+						c.logger.Error("Checking pprof path failed", "error", err)
+						continue
+					}
+					dir = path
+				} else {
+					dir, err = os.MkdirTemp("", "vault-pprof")
+					if err != nil {
+						c.logger.Error("Could not create temporary directory for pprof", "error", err)
+						continue
+					}
+				}
+
+				fGoRoutine, err := os.Create(filepath.Join(dir, "goroutines"))
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				err = pprof.Lookup("goroutine").WriteTo(fGoRoutine, 0)
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				fHeap, err := os.Create(filepath.Join(dir, "heap"))
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				err = pprof.Lookup("heap").WriteTo(fHeap, 0)
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				fAllocs, err := os.Create(filepath.Join(dir, "allocs"))
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				err = pprof.Lookup("allocs").WriteTo(fAllocs, 0)
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				fThread, err := os.Create(filepath.Join(dir, "threadcreate"))
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				err = pprof.Lookup("threadcreate").WriteTo(fThread, 0)
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
```

Comment:
You might be able to shorten some of this logic via something like:

```go
dumps := []string{"goroutines", "heap", "allocs", "threadcreate"}
for _, dump := range dumps {
  pFile, err := os.Create(filepath.Join(dir, dump))
  if err != nil {
    c.logger.Error("error creating pprof file", "name", dump, "error", err)
    continue
  }

  err = pprof.Lookup(dump).WriteTo(pFile, 0)
  if err != nil {
    c.logger.Error("error generating pprof data", "name", dump, "error", err)
    continue
  }
}
```

Reply:
I just realized that my suggestion likely won't work as written because the `continue` will continue the inner for loop. You'd need a break instead probably.

Reply:
I think the continue would be fine but breaking is also OK. Either way you might not get every pprof output.

Reply:
Right, either way it's no different from the previous code where the first error means none of the rest of the pprofs are generated.